### PR TITLE
create Base minimizer and API

### PR DESCRIPTION
**Pull request by mayou36**
WIP!

- MinuitMinimizer works already for minimization
- MinimizerInterface is implemented (`abc.abstractmethod` is mandatory to override)
- BaseMinimizer takes care of the public functions and allows for simple minimizer subclassing
- TF optimizer getting wrapped instead of subclassed

@apuignav @rsilvaco is this what you would have expected?

### CODE REVIEW

File: zfit/core/minimizer.py [hunk 1]
```diff
@@ -5,66 +5,214 @@
...
+
+    @property
+    def edm(self):
+        return self.get_state(copy=False).edm
+
+    @property
+    def fmin(self):
+        return self.get_state(copy=False).fmin
+
+    def hesse(self):
```
Comment (apuignav) [inline zfit/core/minimizer.py]: I guess we can't just remove it? It would break with `abc`?

Comment (mayou36) [reply zfit/core/minimizer.py]: "yes", as long as it is not overriden by the subclass. But why not provide a default implementation?
It's on the TODO ;)

Comment (apuignav) [reply zfit/core/minimizer.py]: It's a philosophy thing: what does a "default" implementation mean? Each minimizer is going to be different, isn't it?

Comment (apuignav) [reply zfit/core/minimizer.py]: Mmm, to be honest, I'm not sure. Maybe let's leave it like it is for now, then we can revisit if the need appears.

Comment (mayou36) [reply zfit/core/minimizer.py]: > like now
-> with a default implementation?

Comment (apuignav) [reply zfit/core/minimizer.py]: Yeah, or if nothing sensible can be made, raise a `NotImplementedError`.

**Comment by apuignav** [comment]
I am not sure I understand these step-wise minimisers: can you clarify?

**Comment by mayou36** [comment]
Sure: it is what is default in TensorFlow optimizers. `step` does one step of the minimization. And e.g. the TF.Optimizers do this.

It's just another way of defining a minimizer. Advantage: you don't have to care about the termination (if you implement just `step`)

**Comment by mayou36** [comment]
Yes, that's the idea. And if you provide _minimize, that's even better

**Comment by apuignav** [comment]
Cool, I'm happy with this, except all the "WIP" part 😉. Do you want me to approve and you do the merge, or you want to clean that up first?

**Comment by mayou36** [comment]
First clean up. Then I'll remove the WIP and ping you again ;)

**Comment by mayou36** [comment]
Looks good from my side, merge?

**Pull request merged**